### PR TITLE
[String] Comparison Speedups

### PR DESCRIPTION
**Pull request by milseman**
A number of improvements to constant-factor costs in String comparison.

Fixes https://bugs.swift.org/browse/SR-9212, and most of https://bugs.swift.org/browse/SR-9270

### CODE REVIEW

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please benchmark

**Comment by milseman** [comment]
Good news is that the benchmarks that improved are much more realistic than the ones that regressed, which are mostly finicky about inlinability. I'll try to investigate further to see if there's a way to reduce the impact on them, but worse case scenario we'd want to take this.

**Comment by milseman** [comment]
Those benchmarks are always comparing strings that are binary-equal, and the old code had a redundant fast-path based on memcmp. That redundancy would cause a perf hit for strings that don't happen to be equal (double-traversal), but these artificial benchmarks improved because of memcmp's vectorization.

I added a tweak that gets us some minor improvement, but we won't be meeting the old numbers on master until we have vectorization.

Still, we are significantly faster than 4.2 on those benchmarks, even if this shows a relative regression from before this PR. This is worth taking.

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please benchmark

File: stdlib/public/core/StringComparison.swift [hunk 1]
```diff
@@ -12,153 +12,244 @@
...
+
+  let left = _StringGutsSlice(_StringGuts(leftUTF8, isASCII: false))
+  let right = _StringGutsSlice(_StringGuts(rightUTF8, isASCII: false))
+  return left.compare(with: right, expecting: expecting)
+}
+
+// Return the point of binary divergence. If they have no binary difference
+// (even if one is longer), returns nil.
+@_effects(readonly)
+private func _findDiffIdx(
```
Comment (Catfish-Man) [inline stdlib/public/core/StringComparison.swift]: Isn't this just strcspn?

Comment (milseman) [reply stdlib/public/core/StringComparison.swift]: Isn't `strcspn` looking for members in a set of characters?

Comment (Catfish-Man) [reply stdlib/public/core/StringComparison.swift]: Interestingly, almost every bit of documentation I can find is ambiguous on this, but yes, I think you're right. I coulda sworn there was a standard version of this, but I'm blanking on it currently. Anyway, carry on! Sorry about the noise.

Comment (jrose-apple) [reply stdlib/public/core/StringComparison.swift]: It's [`std::mismatch`](https://en.cppreference.com/w/cpp/algorithm/mismatch), but you're not going to get that from Swift.

**Pull request merged**